File: typescript/src/application/handlers/TimelineQueryHandler.ts
import type { Handler } from '../Handler'
import type { UserId } from '../../domain/aggregates/User'
import type { PostId } from '../../domain/entities/Post'
import type { UserRepository } from '../repositories/UserRepository'

export class TimelineUserNotFound extends Error {}

type TimelineQuery = { userId: UserId }

type TimelinePost = { postId: PostId; userId: UserId; text: string; dateTime: string }

export class TimelineQueryHandler implements Handler<TimelineQuery, TimelinePost[]> {
  constructor(private readonly userRepository: UserRepository) {}

  async handle({ userId }: TimelineQuery): Promise<TimelinePost[]> {
    const user = await this.userRepository.get(userId)
    if (!user) throw new TimelineUserNotFound()
    return user.posts.map((p) => ({ ...p, userId }))
  }
}

File: scala/src/main/scala/openchat/application/handlers/TimelineQueryHandler.scala
package openchat.application.handlers

import openchat.application.Handler
import openchat.application.repositories.UserRepository

import scala.concurrent.{ExecutionContext, Future}

final case class TimelinePostView(postId: String, userId: String, text: String, dateTime: String)
final class TimelineUserNotFound extends RuntimeException

final class TimelineQueryHandler(userRepository: UserRepository)(implicit ec: ExecutionContext)
    extends Handler[String, Vector[TimelinePostView]] {
  def handle(userId: String): Future[Vector[TimelinePostView]] = {
    for {
      maybeUser <- userRepository.get(userId)
      user      <- maybeUser match {
        case None        => Future.failed(new TimelineUserNotFound)
        case Some(user0) => Future.successful(user0)
      }
    } yield user.posts.map(p =>
      TimelinePostView(postId = p.postId, userId = userId, text = p.text, dateTime = p.dateTime)
    )
  }
}
